File: html5/src/rendering/SkyGradient.js
// -----------------------------------------------------------------------------
// SkyGradient — the colour of a day, interpolated from the table both engines read.
//
// The mirror of SexyBiscuit.Engine/Rendering/SkyGradient.cs.
//
// A day/night cycle is almost entirely colour, and colour written twice is two
// different dusks -- the desktop build warm and the web build blue, with nothing able
// to see it. So the keys live in one JSON file and a shared fixture pins what both
// engines make of them.
//
// Time runs 0 to 1 with 0 at midnight and 0.5 at noon, and the table wraps: the value
// at 0.99 blends towards the key at 0, not off the end of the array.
// -----------------------------------------------------------------------------

import table from './sky-gradient.json' with { type: 'json' };

/** How many keys the table holds. */
export const KEY_COUNT = table.keys.length;

/** The name of each key, in order, for an editor's timeline. */
export const KEY_NAMES = table.keys.map((k) => k.name ?? '');

/** Folds any time into 0..1, so a clock that has run for days still works. */
export function wrap01(t) {
    const wrapped = t % 1;
    return wrapped < 0 ? wrapped + 1 : wrapped;
}

/** The sky at a moment. Any time is accepted; it is wrapped into one day first. */
export function sample(time01) {
    const keys = table.keys;
    if (keys.length === 0) return null;

    const t = wrap01(time01);

    // The last key before t, and the one after it. The table wraps, so "after the last
    // key" is the first key again -- which is what makes 23:59 blend into midnight
    // rather than snapping.
    let before = 0;
    for (let i = 0; i < keys.length; i++) {
        if (keys[i].time > t) break;
        before = i;
    }

    const after = Math.min(before + 1, keys.length - 1);
    const a = keys[before];
    const b = keys[after];

    const span = b.time - a.time;
    const f = span <= 0 ? 0 : Math.min(1, Math.max(0, (t - a.time) / span));

    return {
        zenith: mixHex(a.zenith, b.zenith, f),
        horizon: mixHex(a.horizon, b.horizon, f),
        sun: mixHex(a.sun, b.sun, f),
        sunIntensity: lerp(a.sunIntensity, b.sunIntensity, f),
        ambientSky: mixHex(a.ambientSky, b.ambientSky, f),
        ambientGround: mixHex(a.ambientGround, b.ambientGround, f),
        fog: mixHex(a.fog, b.fog, f),
        fogDensity: lerp(a.fogDensity, b.fogDensity, f),
    };
}

// -----------------------------------------------------------------------------
// Colour
// -----------------------------------------------------------------------------

function lerp(a, b, f) { return a + (b - a) * f; }

/**
 * Blends two hex colours per channel.
 *
 * The rounding rule is stated rather than borrowed, and both engines state the same
 * one: floor(x + 0.5), computed in doubles. MonoGame's Color.Lerp truncates a
 * single-precision lerp and JavaScript's Math.round rounds half away from zero while
 * C#'s MathF.Round rounds half to even -- three different answers for the same blend,
 * each of them off by one byte at some fractions and correct at most.
 */
export function mixHex(a, b, f) {
    const ca = parseHex(a);
    const cb = parseHex(b);
    const amount = Math.min(1, Math.max(0, f));

    const mix = (x, y) => Math.min(255, Math.max(0, Math.floor(x + (y - x) * amount + 0.5)));
    return toHex(mix(ca[0], cb[0]), mix(ca[1], cb[1]), mix(ca[2], cb[2]));
}

function parseHex(hex) {
    let body = String(hex).replace(/^#/, '');
    if (body.length === 3) body = body.split('').map((c) => c + c).join('');
    return [0, 2, 4].map((i) => parseInt(body.slice(i, i + 2), 16));
}

function toHex(r, g, b) {
    return `#${[r, g, b].map((c) => c.toString(16).padStart(2, '0')).join('')}`;
}

File: SexyBiscuit.Engine/Rendering/SkyGradient.cs
using System.Text.Json;
using Microsoft.Xna.Framework;

namespace SexyBiscuit.Engine.Rendering;

/// <summary>
/// Everything the sky, the sun and the ambient light look like at one moment.
/// </summary>
public readonly record struct SkySample(
    Color Zenith,
    Color Horizon,
    Color Sun,
    float SunIntensity,
    Color AmbientSky,
    Color AmbientGround,
    Color Fog,
    float FogDensity);

/// <summary>
/// The colour of a day, interpolated from the table both engines read.
/// </summary>
/// <remarks>
/// <para>
/// The mirror of <c>html5/src/rendering/SkyGradient.js</c>. A day/night cycle is almost
/// entirely colour, and colour written twice is two different dusks — the desktop build
/// warm and the web build blue, with nothing able to see it. So the keys live in one JSON
/// file, embedded here the same way the presets and the glyph table are, and a shared
/// fixture pins what both engines make of them.
/// </para>
/// <para>
/// Time runs 0 to 1 with 0 at midnight and 0.5 at noon, and the table wraps: the value at
/// 0.99 blends towards the key at 0, not off the end of the array.
/// </para>
/// </remarks>
public sealed class SkyGradient
{
    /// <summary>The table every caller uses. Parsed once.</summary>
    public static SkyGradient Shared { get; } = Load();

    private readonly List<Key> _keys = new();

    /// <summary>How many keys the table holds.</summary>
    public int KeyCount => _keys.Count;

    /// <summary>The name of each key, in order, for an editor's timeline.</summary>
    public IEnumerable<string> KeyNames => _keys.Select(k => k.Name);

    /// <summary>
    /// The sky at a moment. Any time is accepted; it is wrapped into one day first.
    /// </summary>
    public SkySample Sample(float time01)
    {
        if (_keys.Count == 0) return default;

        float t = Wrap01(time01);

        // The last key before t, and the one after it. The table wraps, so "after the
        // last key" is the first key again — which is what makes 23:59 blend into
        // midnight rather than snapping.
        int before = 0;
        for (int i = 0; i < _keys.Count; i++)
        {
            if (_keys[i].Time > t) break;
            before = i;
        }

        int after = Math.Min(before + 1, _keys.Count - 1);
        Key a = _keys[before], b = _keys[after];

        // In double, because the browser engine has no float: doing the blend in single
        // precision here and in double there puts the two a whole byte apart wherever a
        // channel lands near a rounding boundary.
        double span = (double)b.Time - a.Time;
        double f = span <= 0d ? 0d : Math.Clamp(((double)t - a.Time) / span, 0d, 1d);

        return new SkySample(
            Zenith:        Mix(a.Zenith, b.Zenith, f),
            Horizon:       Mix(a.Horizon, b.Horizon, f),
            Sun:           Mix(a.Sun, b.Sun, f),
            SunIntensity:  (float)(a.SunIntensity + (b.SunIntensity - a.SunIntensity) * f),
            AmbientSky:    Mix(a.AmbientSky, b.AmbientSky, f),
            AmbientGround: Mix(a.AmbientGround, b.AmbientGround, f),
            Fog:           Mix(a.Fog, b.Fog, f),
            FogDensity:    (float)(a.FogDensity + (b.FogDensity - a.FogDensity) * f));
    }

    /// <summary>
    /// Blends two colours per channel.
    /// </summary>
    /// <remarks>
    /// The rounding rule is stated rather than borrowed, and the browser engine states
    /// the same one: floor(x + 0.5), computed in doubles. MonoGame's
    /// <c>Color.Lerp</c> truncates a single-precision lerp, JavaScript's
    /// <c>Math.round</c> rounds half away from zero and <c>MathF.Round</c> rounds half
    /// to even — three different answers for the same blend, each off by a byte at some
    /// fractions and correct at most.
    /// </remarks>
    private static Color Mix(Color a, Color b, double f)
    {
        double amount = Math.Clamp(f, 0d, 1d);

        static byte Channel(byte x, byte y, double amount)
            => (byte)Math.Clamp(Math.Floor(x + (y - x) * amount + 0.5), 0, 255);

        return new Color(Channel(a.R, b.R, amount),
                         Channel(a.G, b.G, amount),
                         Channel(a.B, b.B, amount),
                         Channel(a.A, b.A, amount));
    }

    /// <summary>Folds any time into 0..1, so a clock that has run for days still works.</summary>
    public static float Wrap01(float t)
    {
        float wrapped = t % 1f;
        return wrapped < 0f ? wrapped + 1f : wrapped;
    }

    // -------------------------------------------------------------------------
    // Parsing
    // -------------------------------------------------------------------------

    private readonly record struct Key(float Time, string Name, Color Zenith, Color Horizon,
                                       Color Sun, float SunIntensity, Color AmbientSky,
                                       Color AmbientGround, Color Fog, float FogDensity);

    private static SkyGradient Load()
    {
        var gradient = new SkyGradient();

        using Stream? stream = typeof(SkyGradient).Assembly
            .GetManifestResourceStream("SexyBiscuit.Engine.sky-gradient.json");

        // An empty table samples to black, which is wrong but is not a crash: a broken
        // build should render something, not refuse to start.
        if (stream is null) return gradient;

        using var document = JsonDocument.Parse(stream);
        foreach (JsonElement key in document.RootElement.GetProperty("keys").EnumerateArray())
        {
            gradient._keys.Add(new Key(
                Time:          key.GetProperty("time").GetSingle(),
                Name:          key.TryGetProperty("name", out JsonElement n) ? n.GetString() ?? "" : "",
                Zenith:        Hex(key, "zenith"),
                Horizon:       Hex(key, "horizon"),
                Sun:           Hex(key, "sun"),
                SunIntensity:  key.GetProperty("sunIntensity").GetSingle(),
                AmbientSky:    Hex(key, "ambientSky"),
                AmbientGround: Hex(key, "ambientGround"),
                Fog:           Hex(key, "fog"),
                FogDensity:    key.GetProperty("fogDensity").GetSingle()));
        }

        return gradient;
    }

    /// <summary>Through the document codec, so one hex parser serves the whole engine.</summary>
    private static Color Hex(JsonElement key, string name)
        => UI.UiDocument.ParseColour(key.GetProperty(name).GetString()) ?? Color.Black;
}
